source/agora/consensus/data/UtxoSet.d

```diff
@@ -0,0 +1,292 @@
+    private Set!Hash used_utxos;
+
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            utxo_db_path = path to the UTXO database
+
+    ***************************************************************************/
+
+    public this (in string utxo_db_path)
+    {
+        this.utxo_db = new UTXODB(utxo_db_path);
+    }
+
+    /***************************************************************************
+
+        Shut down the utxo store
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.utxo_db.shutdown();
+    }
+
+    /***************************************************************************
+
+        Add all of a transaction's outputs to the Utxo set,
+        and remove the spent outputs in the transaction from the set.
+
+        Params:
+            tx = the transaction
+
+    ***************************************************************************/
+
+    public void updateUtxoCache (const ref Transaction tx) nothrow @safe
+    {
+        foreach (input; tx.inputs)
+        {
+            auto utxo_hash = getHash(input.previous, input.index);
+            this.utxo_db.remove(utxo_hash);
+        }
+
+        Hash tx_hash = tx.hashFull();
+        foreach (idx, output; tx.outputs)
+        {
+            auto utxo_hash = getHash(tx_hash, idx);
+            this.utxo_db[utxo_hash] = output;
+        }
+    }
+
+    /***************************************************************************
+
+        Reset the tracking of spent UTXOs
+
+    ***************************************************************************/
+
+    public void reset () @trusted nothrow
+    {
+        this.used_utxos.clear();
+    }
+
+    /***************************************************************************
+
+        Find an unspent Output in the UTXO set.
+
+        Params:
+            tx_hash = the hash of transation
+            index = the index of the output
+            output = will contain the UTXO if found
+
+        Return:
+            Return true if the UTXO was found
+
+    ***************************************************************************/
+
+    public bool findUTXO (Hash hash, size_t index, out Output output)
+        nothrow @safe
+    {
+        auto utxo_hash = getHash(hash, index);
+
+        if (utxo_hash in this.used_utxos)
+            return false;  // double-spend
+
+        if (this.utxo_db.find(utxo_hash, output))
+        {
+            this.used_utxos.put(utxo_hash);
+            return true;
+        }
+
+        return false;
+    }
+
+    /***************************************************************************
+
+        Get the combined hash of the previous hash and index.
+        This makes sure the index is always of the same type,
+        as mixing different-sized uint/ulong would create different hashes.
```

Comment:
You can just use `ulong(...)` at call site though, but I don't mind if we have a function for it.

Reply:
I originally wrote it like that, but then I realized I had three call sites where I'm casting to `ulong`, and I want to avoid adding duplicate comments at the call site explaining why it's done this way.

Also, tracking down the hashing "bug" took a good portion of my time.